lib/Serialization/Serialization.cpp

```diff
@@ -3284,6 +3284,7 @@ void Serializer::writeDecl(const Decl *D) {
                                rawAccessLevel,
                                inherited);
 
+    const_cast<ProtocolDecl*>(proto)->createGenericParamsIfMissing();
     writeGenericParams(proto->getGenericParams());
```

Comment:
In a separate PR can you try moving the createGenericParamsIfMissing() call into getGenericParams() itself?